lib/Frontend/CompilerInvocation.cpp

```diff
@@ -435,18 +435,28 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
       return true;
     }
   } else if (Opts.RequestedAction != FrontendOptions::NoneAction) {
-    if (Opts.InputFilenames.empty()) {
+    if (Opts.InputFilenames.empty() && !Args.hasArg(OPT_e)) {
       Diags.diagnose(SourceLoc(), diag::error_mode_requires_an_input_file);
       return true;
     }
   }
 
   if (Opts.RequestedAction == FrontendOptions::Immediate) {
-    assert(!Opts.InputFilenames.empty());
-    Opts.ImmediateArgv.push_back(Opts.InputFilenames[0]); // argv[0]
-    if (const Arg *A = Args.getLastArg(OPT__DASH_DASH)) {
-      for (unsigned i = 0, e = A->getNumValues(); i != e; ++i) {
-        Opts.ImmediateArgv.push_back(A->getValue(i));
+    if (Args.hasArg(OPT_e)) {
+      for (const Arg *A : Args.filtered(OPT_e)) {
+        if (A->getNumValues() > 0) {
+          auto sourceBuffer =
+            llvm::MemoryBuffer::getMemBufferCopy(A->getValue(), "immediate");
+          Opts.InputBuffers.push_back(std::move(sourceBuffer.get()));
+        }
```

Comment:
Still crashing down the line as a result of this, but I'm not understanding what I'm doing wrong here.

Reply:
I am not that familiar with this part of the code, but at a surface level it appears that sourceBuffer is being deallocated at end of scope. This is b/c Opts.InputBuffer only stores a pointer, so you haven't actually moved the "owning" reference into Opts.InputBuffer since Opts.InputBuffer actually doesn't own its membuffers.

I am not sure why the code is designed this way. I think @jrose-apple would know.

Reply:
Yeah, that makes sense. Thanks for the insights on `Opts.InputBuffer`.

Reply:
I suspect it's because the primary use of InputBuffer is for SourceKit, and they probably want their buffer to stay alive *past* the end of the execution. We'll have to get something else to keep the buffer alive. The easiest way to do that is probably to put a `std::unique_ptr<llvm::MemoryBuffer>` member into CompilerInvocation, so that the contents can last as long as the CompilerInvocation does.

Reply:
Sounds good. Thanks for the feedback